src/StrawberryfieldHydroponicsService.php

```diff
@@ -181,64 +297,116 @@ public function run() {
+    }
+    //The normal running condition when all is ok
+    if (($deltaTime <= $heartbeat_max_delta) && ($queuerunner_pid > 0) && $running_posix) {
       $return = [
        'running' => TRUE,
+       'error' => FALSE,
        'message' =>  $this->t('Hydroponics Service Is Running on PID @pid, time passed since last seen @time', [
-        '@time' => ($currentTime - $lastRunTime),
-        '@pid' => $queuerunner_pid,
+        '@time' => $deltaTime,
+        '@pid' => $queuerunner_pid
        ]),
        'PID' => $queuerunner_pid
+      ];
+    }
+    //Normal stopped condition when all is ok
+    elseif (($deltaTime > $heartbeat_max_delta) && ($queuerunner_pid <= 0) && !$running_posix) {
+    //NO MORE NEEDED => Be more conservative so:
+    //Stopped condition without heartbeat check
+    //elseif (($queuerunner_pid <= 0) && !$running_posix) {
+      $return = [
+        'running' => FALSE,
+        'error' => FALSE,
+        'message' =>  $this->t('Hydroponics Service Not running, time passed since last seen @time s', [
+         '@time' => $deltaTime
+        ]),
+        'PID' => $queuerunner_pid,
+        'DELTA' => $deltaTime,
+        'PTABLE' => $running_posix
+      ];
+    }
+    //Something went wrong. See logs and try a reset
+    else {
+     $return = [
+       'running' => FALSE,
+       'error' => TRUE,
+       'message' =>  $this->t('Hydroponics Service ERROR! Check logs and try a reset. Time passed since last seen @time s, last PID @pid, on Process table @ptable', [
+        '@time' => $deltaTime,
+        '@pid' => $queuerunner_pid,
+        '@ptable' => ($running_posix) ? "YES" : "NO"
+       ]),
+       'PID' => $queuerunner_pid,
+       'DELTA' => $deltaTime,
+       'PTABLE' => $running_posix
      ];
     }
     return $return;
   }
 
+  //This function is used for stop and for reset
   public function stop() {
     $queuerunner_pid = (int) \Drupal::state()->get('hydroponics.queurunner_last_pid', 0);
     $lastRunTime = intval(\Drupal::state()->get('hydroponics.heartbeat'));
     $currentTime = intval(\Drupal::time()->getRequestTime());
-    error_log($queuerunner_pid);
-    $running_posix = posix_kill($queuerunner_pid, 0);
-    if (!$running_posix || !$queuerunner_pid) {
+    $deltaTime = ($currentTime - $lastRunTime);
+
+    $heartbeat_max_delta = $this->getHearbeatMaxDelta();
+
+    $running_posix = FALSE;
+    if ($queuerunner_pid > 0) {
+      $running_posix = posix_kill($queuerunner_pid, 0);
+    }
+    //Normal stopped condition when all is ok
+    if (($deltaTime > $heartbeat_max_delta) && ($queuerunner_pid <= 0) && !$running_posix) {
       return NULL;
-    } else {
+    }
+    //process in table and pid available regardless of deltaTime
+    elseif (abs($queuerunner_pid) > 0 && $running_posix) {
       if (extension_loaded('pcntl')) {
-        $running_posix = posix_kill($queuerunner_pid, SIGTERM);
+        $running_posix = posix_kill(abs($queuerunner_pid), SIGTERM);
```

Comment:
We may want to have a comment somewhere like here https://github.com/esmero/strawberryfield/pull/138/files#diff-f87fbb18e8f039828bd65f2204e85b8a323c7f253f13ea2f5109174060aa3a4eR388 (every time we fetch the PID) explaining that if its negative means it was started? basically explain why you use negatives for tracking PIDS.

Reply:
Added a more verbose checkRunning result, see if what you mean or not, please